### PR TITLE
[XLA] Make sure HloComputation::Equals takes dangling side-effects into account.

### PR DESCRIPTION
**Pull request by georgepaw**
Current implementation does not take dangling side-effect instructions such as outfeeds or rngs into account.
For example:
```
c1 (arg0.1: f32[], arg1.2: f32[2], arg2.3: f32[1,1,2,2]) -> () {
  %arg0.1 = f32[] parameter(0)
  %arg2.3 = f32[1,1,2,2]{3,2,1,0} parameter(2)
  %arg1.2 = f32[2]{0} parameter(1)
  // Do some computation
  %result = fp32[] ....
  %aa = token[] after-all()
  %feed = token[] outfeed(result.35, aa)
  ROOT %tuple.53 = () tuple()
}

c2 (arg0.1: f32[], arg1.2: f32[2], arg2.3: f32[1,1,2,2]) -> () {
  %arg0.1 = f32[] parameter(0)
  %arg2.3 = f32[1,1,2,2]{3,2,1,0} parameter(2)
  %arg1.2 = f32[2]{0} parameter(1)
  ROOT %tuple.53 = () tuple()
}
```
then `*c1 == *c2` evaluates to true.

### CODE REVIEW

**Comment by georgepaw** [comment]
Note that we could simplify this by saying "two computations can never be equal if they have side effects". And it seems like most instructions with side-effect override identical to `false` but there are exceptions like the all-reduce in this test.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/hlo_computation_test.cc [hunk 1]
```diff
@@ -703,5 +703,104 @@ ENTRY entry {
...
+  EXPECT_FALSE(*comp4 == *comp0);
+
+  EXPECT_TRUE(*comp1 == *comp2);
+  EXPECT_TRUE(*comp2 == *comp1);
+  EXPECT_FALSE(*comp1 == *comp3);
+  EXPECT_FALSE(*comp3 == *comp1);
+  EXPECT_FALSE(*comp1 == *comp4);
+  EXPECT_FALSE(*comp4 == *comp1);
+
+  EXPECT_FALSE(*comp2 == *comp3);
```
[the remark this answers is not in the document]
Comment (georgepaw) [reply tensorflow/compiler/xla/service/hlo_computation_test.cc]: Sure, I was just following other tests in the file

File: tensorflow/compiler/xla/service/hlo_computation.cc [hunk 2]
```diff
@@ -772,34 +772,97 @@ bool HloComputation::Equal(const HloComputation& other,
...
+  if (side_effect_instrs_this.size() != side_effect_instrs_other.size()) {
+    return false;
+  }
+
+  // Make sure there is a mapping for each side-effect instruction.
+  for (const HloInstruction* instr_this : side_effect_instrs_this) {
+    bool found_match = false;
+    for (const HloInstruction* instr_other : side_effect_instrs_other) {
+      if (identical_from_instr(instr_this, instr_other)) {
+        // Make sure an instruction is only matched once.
```
[the remark this answers is not in the document]
Comment (georgepaw) [reply tensorflow/compiler/xla/service/hlo_computation.cc]: Don't really follow this one, did you mean
> `instr_this` matches **two** instructions in `side_effect_instrs_other`
?
For every instruction with side-effect in "this" we find an instruction in "other" which matches. If there are two instructions in "other" which match an instruction from "this" we would expect two instructions in "this" as well for the computations to be identical.
We could check the other set is empty at the end, but we make sure we have the same number of side effect instructions before, and for every instruction in "this" we either remove an instruction from "other" if we match or we return early if we don't so the set should implicitly be empty.

File: tensorflow/compiler/xla/service/hlo_computation.cc [hunk 3]
```diff
@@ -772,34 +772,97 @@ bool HloComputation::Equal(const HloComputation& other,
...
     return true;
   }
+
+  // Keep track of the pairs of instructions which have been matched.
   absl::flat_hash_set<std::pair<const HloInstruction*, const HloInstruction*>>
-      visited;
-  std::vector<std::pair<const HloInstruction*, const HloInstruction*>> worklist;
+      matched_between_comps;
+
+  auto identical_from_instr = [&matched_between_comps, is_layout_sensitive](
```
[the remark this answers is not in the document]
Comment (georgepaw) [reply tensorflow/compiler/xla/service/hlo_computation.cc]: Sure, should it be in an anonymous namespace? It also uses an internal state `matched_between_comps` to speed up the lookup for dangling instructions - do we want to keep that?

**Comment by gbaned** [comment]
@georgepaw Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@georgepaw gentle ping to resolve conflicts? Thanks!

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**